### PR TITLE
[GEARPUMP-361] Build gearpump docker image locally

### PR DESCRIPTION
**Pull request by manuzhang**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the commit message is formatted like:
   `[GEARPUMP-<Jira issue #>] Meaningful description of pull request` 
 - [ ] Make sure tests pass via `sbt clean test`.
 - [ ] Make sure old documentation affected by the pull request has been updated and new documentation added for new functionality.

### CODE REVIEW

**Comment by manuzhang** [comment]
R: @karol-brejna-i @huafengw Build Gearpump Image locally to make it easier to update and fix bugs

**Comment by huafengw** [comment]
Verified on centos, the updated image can be launched successfully.

**Comment by karol-brejna-i** [comment]
Even when providing proxy args for docker build:
```
docker build --no-cache --build-arg HTTP_PROXY=$http_proxy --build-arg HTTPS_PROXY=$https_proxy -t gearpump/gearpump-launcher integrationtest/docker/gearpump
```

It still fails to install python. I'll try to find a workaround for this.

**Comment by karol-brejna-i** [comment]
I've just confirmed. The problem persists, even when networking/proxy problems are overcomed:

```
$ docker build --no-cache -t gearpump/gearpump-launcher integrationtest/docker/gearpump
Sending build context to Docker daemon  8.192kB
Step 1/7 : FROM grubykarol/jre:8u121b13_server
 ---> f7e482e81614
Step 2/7 : RUN opkg-install python
 ---> Running in 0939e0c36efc
Downloading http://downloads.openwrt.org/snapshots/trunk/x86/64/packages/base/Packages.gz.
Inflating http://downloads.openwrt.org/snapshots/trunk/x86/64/packages/base/Packages.gz.
Updated list of available packages in /var/opkg-lists/base.
Downloading http://downloads.openwrt.org/snapshots/trunk/x86/64/packages/packages/Packages.gz.
Inflating http://downloads.openwrt.org/snapshots/trunk/x86/64/packages/packages/Packages.gz.
Updated list of available packages in /var/opkg-lists/packages.
Unknown package 'python'.
Collected errors:
 * opkg_install_cmd: Cannot install package python.
```


`Unknown package 'python'.`

**Comment by manuzhang** [comment]
It seems python package is not supported by busybox. Can we switch to the official `openjdk/8-jre-alpine` image ? It's 120MB after installing python.

**Comment by karol-brejna-i** [comment]
Docker image problem solved with https://github.com/apache/incubator-gearpump/pull/235.

**Comment by manuzhang** [comment]
@karol-brejna-i @huafengw rebased. Please help to verify again

**Comment by huafengw** [comment]
1. I have to build with explicit proxies like 
`docker build --build-arg HTTP_PROXY=... --build-arg HTTPS_PROXY=...`.
If not,  building step
`fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/main/x86_64/APKINDEX.tar.gz` will always fail.

2. Please add following change in `integrationtest/docker/gearpump/Dockerfile`
```
-RUN apk add --no-cache python && \
+RUN apk add --update curl && \
+    apk add --no-cache python && \
```

3. After resolved these issues, the integration test still can't proceed. Here is the error log:
```
[DEBUG] [15:21:49.279] [Docker$] Container master0====>> curl http://master0:8090/api/v1.0/master
[DEBUG] [15:21:49.279] [Docker$] INPUT==>>
[DEBUG] [15:21:49.279] [ShellExec$] EXEC master0 => `docker exec master0 curl -s  --cookie cookie.txt http://master0:8090/api/v1.0/master`
[DEBUG] [15:21:49.636] [ShellExec$] EXEC master0 <= `<HTML>
<HEAD><TITLE>Redirection</TITLE></HEAD>
<BODY><H1>Redirect</H1></BODY>
</HTML>` exit 0
[DEBUG] [15:21:49.636] [Docker$] <<==OUTPUT
[DEBUG] [15:21:49.636] [Docker$] <HTML>
<HEAD><TITLE>Redirection</TITLE></HEAD>
<BODY><H1>Redirect</H1></BODY>
</HTML>
[DEBUG] [15:21:49.637] [Docker$] <<====Command END. Container master0, curl http://master0:8090/api/v1.0/master

[ERROR] [15:21:49.637] [RestClient] Failed to decode Rest response to MasterData
[ERROR] [15:21:49.637] [Util$] Failed due to (false == cluster running), retrying for the 15 times...
```

**Comment by huafengw** [comment]
The 3rd problem occurs on Mac and centos works fine.

**Comment by karol-brejna-i** [comment]
The third problem - works for me on Ubuntu 16.04. 

```
[DEBUG] [05:25:55.233] [Docker$] Container master0====>> curl http://master0:8090/api/v1.0/master
[DEBUG] [05:25:55.233] [Docker$] INPUT==>>
[DEBUG] [05:25:55.233] [ShellExec$] EXEC master0 => `docker exec master0 curl -s  --cookie cookie.txt http://master0:8090/api/v1.0/master`
[DEBUG] [05:25:55.379] [ShellExec$] EXEC master0 <= `{"masterDescription":{"leader":{"host":"master@master0","port":3000},"cluster":[{"host":"master0","port":3000}],"aliveFor":"8068","logFile":"/opt/gearpump/var/log/gearpump/daemon","jarStore":"jarstore...` exit 0
[DEBUG] [05:25:55.379] [Docker$] <<==OUTPUT
```
Checking on Mac...

**Comment by manuzhang** [comment]
@huafengw can't reproduce 1 and 3 on Mac

File: integrationtest/docker/gearpump/Dockerfile [hunk 1]
```diff
@@ -18,7 +18,8 @@
 FROM openjdk:8u131-jre-alpine
 
 # Install python
-RUN apk add --no-cache python && \
+RUN apk add --update curl && \
```
Comment (karol-brejna-i) [inline integrationtest/docker/gearpump/Dockerfile]: please, use `apk add --update curl python` (one line)

**Comment by karol-brejna-i** [comment]
```[DEBUG] [13:07:44.774] [Docker$] Container master0====>> curl http://master0:8090/api/v1.0/master
[DEBUG] [13:07:44.774] [Docker$] INPUT==>>
[DEBUG] [13:07:44.775] [ShellExec$] EXEC master0 => `docker exec master0 curl -s  --cookie cookie.txt http://master0:8090/api/v1.0/master`
[DEBUG] [13:07:45.490] [ShellExec$] EXEC master0 <= `{"masterDescription":{"leader":{"host":"master@master0","port":3000},"cluster":[{"host":"master0","port":3000}],"aliveFor":"18124","logFile":"/opt/gearpump/logs","jarStore":"jarstore/","masterStatus":...` exit 0
[DEBUG] [13:07:45.490] [Docker$] <<==OUTPUT
[DEBUG] [13:07:45.490] [Docker$] {"masterDescription":{"leader":{"host":"master@master0","port":3000},"cluster":[{"host":"master0","port":3000}],"aliveFor":"18124","logFile":"/opt/gearpump/logs","jarStore":"jarstore/","masterStatus":"synced","homeDirectory":"/opt/gearpump","activities":[],"jvmName":"7@master0","historyMetricsConfig":{"retainHistoryDataHours":72,"retainHistoryDataIntervalMs":3600000,"retainRecentDataSeconds":300,"retainRecentDataIntervalMs":15000}}}
[DEBUG] [13:07:45.491] [Docker$] <<====Command END. Container master0, curl http://master0:8090/api/v1.0/master 
```

Number 3 looks good on mac, too. For me it seems like some proxy-caused redirect...

**Comment by huafengw** [comment]
+1

**Pull request closed without merging**